mismi-s3/src/Mismi/S3/Commands.hs

```diff
@@ -280,10 +281,23 @@ listObjects :: Address -> AWS ([Address], [Address])
 listObjects a =
   (\(p, k) -> (Address (bucket a) <$> p, Address (bucket a) <$> k) )<$> getObjects a
 
--- list the addresses, keys first, then prefixes
 list :: Address -> AWS [Address]
 list a =
-  (\(p, k) -> k <> p) <$> listObjects a
+  list' a >>= ($$ DC.consume)
+
+list' :: Address -> AWS (Source AWS Address)
+list' a@(Address (Bucket b) (Key k)) = do
+  let pp kk = if T.null kk then "" else if T.isSuffixOf "/" kk then kk else kk <> "/"
```

Comment:
Is this different to `<>` if so, should `<>` change? Either way probably good to have a test of some-sort to sanity check this.